### PR TITLE
feat: Provide mechanism for dynamic config setup at app initialization phase

### LINKED ISSUE #5181: Provide mechanism for dynamic config setup at app initialization phase
Currently, the config is provided statically as an injection token and should not be changed at runtime. 

There are cases, however, where allowing to change config at the app initialization phase would greatly simplify the implementation of selected features, making it justified. 

Blocks #5164

### PR DESCRIPTION
**Pull request by dunqan**
Currently, the config is provided statically as an injection token and should not be changed at runtime.

There are cases, however, where allowing to change config at the app initialization phase would greatly simplify the implementation of selected features, making it justified.

Closes #5181

### CODE REVIEW

File: projects/core/src/config/config-initializer/config-initializer.service.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      if (!this.isReady(ongoing, waitingScope)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  protected isReady(ongoing: string[], waitingScope: string): boolean {
+    for (const scope of ongoing) {
+      if (waitingScope.includes(scope)) {
```
Comment (Platonn) [inline projects/core/src/config/config-initializer/config-initializer.service.ts]: 1. It matches also strings in the middle, i.e. `'context'.includes('on')`
2. `startsWith` won't suffice too. We should split scopes with dots, not to match prefix of strings:
`'context.baseSite'.includes('context.base')`

File: projects/core/src/config/config-initializer/config-initializer.module.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+@NgModule({
+  declarations: [],
+  imports: [CommonModule],
+})
+export class ConfigInitializerModule {
+  static forRoot(): ModuleWithProviders<ConfigInitializerModule> {
+    return {
+      ngModule: ConfigInitializerModule,
+      providers: [
+        {
```
Comment (Platonn) [inline projects/core/src/config/config-initializer/config-initializer.module.ts]: What about `provideConfig({ initializing: true })`?

File: projects/core/src/config/config-initializer/config-initializer.service.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      return;
+    }
+
+    const ongoingScopes = ['initializing'];
+
+    const asyncConfigs = [];
+
+    for (const initializer of this.initializers) {
+      if (isDevMode() && !(initializer.scopes && initializer.scopes.length)) {
+        console.error('ConfigInitializer should provide scope!');
```
Comment (Platonn) [inline projects/core/src/config/config-initializer/config-initializer.service.ts]: What about replacing `ConfigInitializer` to the name of the token `CONFIG_INITIALIZER` in both messages - for discoverability?

File: projects/core/src/config/config-initializer/config-initializer.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { InjectionToken } from '@angular/core';
+
+export const CONFIG_INITIALIZER = new InjectionToken('ConfigInitializer');
+
+export interface ConfigInitializer {
+  scopes: string[];
+  configFactory: () => {};
```
Comment (Platonn) [inline projects/core/src/config/config-initializer/config-initializer.ts]: What about simply `factory`?
What about type `() => Promise<any> | void` - for self-documenting?

File: projects/core/src/config/config-initializer/config-initializer.service.ts (hunk 1 above)
Comment (dunqan) [reply projects/core/src/config/config-initializer/config-initializer.service.ts]: thanks!

File: projects/core/src/config/config-initializer/config-initializer.module.ts (hunk 2 above)
Comment (dunqan) [reply projects/core/src/config/config-initializer/config-initializer.module.ts]: 👍

File: projects/core/src/config/config-initializer/config-initializer.service.ts (hunk 3 above)
Comment (dunqan) [reply projects/core/src/config/config-initializer/config-initializer.service.ts]: 👍

File: projects/core/src/config/config-initializer/config-initializer.ts (hunk 4 above)
Comment (dunqan) [reply projects/core/src/config/config-initializer/config-initializer.ts]: Agree with `Promise<any>` 👍 
Any specific reason for 'void'? 
Shouldn't config initializer always provide config chunk?

Comment (dunqan) [reply projects/core/src/config/config-initializer/config-initializer.ts]: 2. `configFactory` is a little bit more obvious to me (a factory that should create config)

**Pull request merged**